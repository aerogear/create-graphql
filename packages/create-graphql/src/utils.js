import shell from 'shelljs';
import chalk from 'chalk';
import ora from 'ora';
import spawn from 'cross-spawn-promise';
import updateNotifier from 'update-notifier';

import pkg from '../package.json';

const tic = chalk.green('✓');
const tac = chalk.red('✗');

const installYeoman = async () => {
  const spinner = ora('Installing Yeoman...');

  spinner.start();

  const command = 'npm';
  const args = ['install', '-g', 'yo'];
  const options = {
    shell: true,
    stdio: false,
  };

  try {
    await spawn(command, args, options);

    spinner.stop();

    console.log(`${tic} Yeoman installed!`);
  } catch (error) {
    spinner.stop();

    console.error(`${tac} There was an error while trying to install Yeoman:`, error);
  }
};

export const verifyYeoman = async () => { // eslint-disable-line import/prefer-default-export
  if (!shell.which('yo')) {
    console.error(`${tac} GraphQL CLI requires Yeoman to be installed.`);

    await installYeoman();
  }

  return true;
};

export const verifyVersion = async () => { // eslint-disable-line import/prefer-default-export
  return await updateNotifier({pkg}).notify();
}
