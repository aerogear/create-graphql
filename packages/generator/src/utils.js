import path from 'path';
import fs from 'fs';
import recast from 'recast';
import pkgDir from 'pkg-dir';

const { visit } = recast.types;

const rootPath = pkgDir.sync('.') || '.';

const getModelCode = modelPath => fs.readFileSync(modelPath, 'utf8');

const parseFieldToGraphQL = (field) => {
  const graphQLField = {
    name: field.name,
    description: field.description,
    required: !!field.required,
  };

  switch (field.type) {
    case 'Number':
      return {
        type: 'GraphQLInt',
        flowType: 'number',
        ...graphQLField,
      };
    case 'Boolean':
      return {
        type: 'GraphQLBoolean',
        flowType: 'boolean',
        ...graphQLField,
      };
    case 'ObjectId': {
      return {
        type: 'GraphQLID',
        flowType: 'string',
        ...graphQLField,
      };
    }
    default:
      return {
        type: 'GraphQLString',
        flowType: 'string',
        ...graphQLField,
      };
  }
};

const parseGraphQLSchema = (mongooseFields) => {
  const dependencies = [];
  const fields = Object.keys(mongooseFields).map((name) => {
    const field = parseFieldToGraphQL(mongooseFields[name]);

    if (dependencies.indexOf(field.type) === -1) {
      dependencies.push(field.type);
    }

    return field;
  });

  return {
    fields,
    dependencies,
  };
};

const getSchemaFieldsFromAst = (node) => {
  const astSchemaFields = node.arguments[0].properties;

  const fields = [];
  astSchemaFields.forEach((field) => {
    const name = field.key.name;
    const fieldDefinition = {};

    if (field.value.type === 'ArrayExpression') {
      return;
    }

    field.value.properties.forEach(({ key, value }) => {
      const fieldName = key.name;

      fieldDefinition[fieldName] = value.name || value.value;
    });

    fields[name] = {
      name,
      ...fieldDefinition,
    };
  });

  return fields;
};

const getSchemaDefinition = (modelCode) => {
  const ast = recast.parse(modelCode, {
    parser: {
      parse: source => require('babylon').parse(source, {
        sourceType: 'module',
        plugins: [
          'asyncFunctions',
          'asyncGenerators',
          'classConstructorCall',
          'classProperties',
          'flow',
          'objectRestSpread',
          'trailingFunctionCommas',
        ],
      }),
    },
  });

  let fields = null;

  visit(ast, {
    visitExpression: function visitExpression(expressionPath) { // eslint-disable-line object-shorthand
      const { node } = expressionPath;

      if (
        node.type === 'NewExpression' &&
        node.callee.object.name === 'mongoose' &&
        node.callee.property.name === 'Schema'
      ) {
        fields = getSchemaFieldsFromAst(node);

        this.abort();
      }

      return this.traverse(expressionPath);
    },
  });

  return parseGraphQLSchema(fields);
};

/**
 * Parse `.graphqlrc` config file and retrieve its contents
 * @param filePath {string} The path of the config file
 * @returns {*}
 */
const parseConfigFile = (filePath) => {
  const config = JSON.parse(fs.readFileSync(filePath, 'utf8'));

  const directories = Object.keys(config.directories).reduce((data, directory) => {
    if (directory === 'source') {
      return {
        ...data,
        [directory]: `${rootPath}/${config.directories[directory]}`,
      };
    }

    return {
      ...data,
      [directory]: `${config.directories.source}/${config.directories[directory]}`,
    };
  }, {});

  return {
    ...config,
    directories: {
      ...config.directories,
      ...directories,
    },
  };
};

/**
 * Get the `.graphqlrc` config file
 * @returns {object} The content of the config
 */
export const getCreateGraphQLConfig = () => {
  // Use default config
  const defaultFilePath = path.resolve(`${__dirname}/graphqlrc.json`);

  const defaultConfig = parseConfigFile(defaultFilePath);

  try {
    // Check if there is a `.graphqlrc` file in the root path
    const customConfig = parseConfigFile(`${rootPath}/.graphqlrc`);

    // If it does, extend default config with it, so if the custom config has a missing line
    // it won't throw errors
    return {
      ...defaultConfig,
      ...customConfig,
    };
  } catch (err) {
    // Return the default config if the custom doesn't exist
    return defaultConfig;
  }
};

/**
 * Get a directory from the configuration file
 * @param directory {string} The name of the directory, e.g. 'source'/'mutation'
 * @returns {string} The directory path
 */
export const getConfigDir = (directory) => getCreateGraphQLConfig().directories[directory];

/**
 * Get the relative path directory between two directories specified on the config file
 * @param from {string} The calling directory of the script
 * @param to {[string]} The destination directories
 * @returns {string} The relative path, e.g. '../../src'
 */
export const getRelativeConfigDir = (from, to) => {
  const config = getCreateGraphQLConfig().directories;

  return to.reduce((directories, dir) => {
    const relativePath = path.relative(config[from], config[dir]);

    return {
      ...directories,
      [dir]: relativePath,
    };
  }, {});
};

export const getMongooseModelSchema = (model) => {
  const modelDir = getCreateGraphQLConfig().directories.model;

  const modelPath = path.resolve(`${modelDir}/${model}.js`);

  const modelCode = getModelCode(modelPath);

  return getSchemaDefinition(modelCode);
};

/**
 * Camel cases text
 * @param text {string} Text to be camel-cased
 * @returns {string} Camel-cased text
 */
export const camelCaseText = (text) => {
  return text.replace(/(?:^\w|[A-Z]|\b\w|\s+)/g, (match, index) => {
    if (+match === 0) {
      return '';
    }

    return index === 0 ? match.toLowerCase() : match.toUpperCase();
  });
};

/**
 * Uppercase the first letter of a text
 * @param text {string}
 * @returns {string}
 */
export const uppercaseFirstLetter = text => `${text.charAt(0).toUpperCase()}${text.slice(1)}`;