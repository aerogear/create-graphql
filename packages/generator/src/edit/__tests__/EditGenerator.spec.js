import helper from 'yeoman-test';
import assert from 'yeoman-assert';
import path from 'path';

import {
  getFileContent,
} from '../../../test/helpers';

import { getConfigDir } from '../../utils';

const editGenerator = path.join(__dirname, '..');

it('generate edit and edit mutation files', async () => {
  const folder = await helper.run(editGenerator)
    .withArguments('Example')
    .toPromise();

  const destinationDir = getConfigDir('edit');

  assert.file([
    `${destinationDir}/ExampleEdit.js`, `${destinationDir}/ExampleEditMutation.js`,
  ]);

  const files = {
    edit: getFileContent(`${folder}/${destinationDir}/ExampleEdit.js`),
    editMutation: getFileContent(`${folder}/${destinationDir}/ExampleEditMutation.js`),
  };

  expect(files).toMatchSnapshot();
});
