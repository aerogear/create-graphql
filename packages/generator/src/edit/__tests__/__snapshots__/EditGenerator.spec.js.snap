// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate edit and edit mutation files 1`] = `
Object {
  "edit": "import React, { Component, PropTypes } from 'react';
import Relay from 'react-relay';
import RelayStore from '../../../RelayStore';
import { withRouter } from 'react-router';

import ExampleEditMutation from './ExampleEditMutation.js';

import Form from '../../common/Form';

class ExampleEdit extends Component {
  static contextTypes = {
    showSnackbar: PropTypes.func,
  };

  fields = [
    {
      name: 'id',
      placeholder: 'ID',
      required: true,
    },
    // TODO - add ObjectType fields here
  ];

  onSubmit = (data) => {
    const { example } = this.props;

    const mutation = new ExampleEditMutation({
      example,
      ...data,
    });

    RelayStore.commitUpdate(mutation, {
      onSuccess: () => {
        this.context.showSnackbar({
          message: 'Example edited successfully!',
        });

        this.props.router.goBack();
      },
      onFailure: (failureResponse) => {
        this.context.showSnackbar({
          message: 'There was an error while trying to edit this Example.',
        });

        console.log('FAIL', failureResponse);
      },
    });
  };

  render() {
    const { example } = this.props;

    return (
      <Form
        fields={this.fields}
        onSubmit={this.onSubmit}
        value={example}
      />
    );
  }
}

const styles = {
  formContainer: {
    display: 'flex',
    flexWrap: 'wrap',
    paddingTop: 30,
    paddingLeft: 10,
  },
  actionsContainer: {
    display: 'flex',
    justifyContent: 'flex-end',
    marginTop: 5,
    paddingRight: 8,
    borderTopStyle: 'solid',
    borderTopWidth: 1,
    paddingTop: 15,
    borderColor: '#ECECEC',
  },
  formField: {
    marginRight: 10,
    flex: '1 0 47%',
  },
  selectField: {
    marginRight: 10,
    flex: '1 0 48%',
  },
};

export default Relay.createContainer(withRouter(ExampleEdit), {
  initialVariables: {
    id: null,
  },
  fragments: {
    example: () => Relay.QL\`
      fragment on Example {
        id
        \${ExampleEditMutation.getFragment('example')}
      }
    \`,
    viewer: () => Relay.QL\`
      fragment on Viewer {
        id
      }
    \`,
  },
});
",
  "editMutation": "import Relay from 'react-relay';

export default class ExampleEditMutation extends Relay.Mutation {
  static fragments = {
    example: () => Relay.QL\`
      fragment on Example {
        id
      }
    \`,
  };

  getMutation() {
    return Relay.QL\`mutation {
      ExampleEdit
    }\`;
  }

  getVariables() {
    const {
      example: {
        id,
      },
      // Todo add more mutation input fields here
    } = this.props;

    return {
      id,
      // TODO add more mutation input fields here
    };
  }

  getFatQuery() {
    return Relay.QL\`
      fragment on ExampleEditPayload {
        example
      }
    \`;
  }

  getConfigs() {
    return [{
      type: 'FIELDS_CHANGE',
      fieldIDs: {
        example: this.props.example.id,
      },
    }];
  }
}
",
}
`;
